nixos/modules/services/development/jupyter.nix

```diff
@@ -0,0 +1,236 @@
+            config = builtins.toJSON {
+              argv = ([ "${kernel.env}/bin/${kernel.executable}" ] ++ kernel.argv);
+              display_name = kernel.name;
+              language = kernel.language;
+            };
+          in ''
+            mkdir kernels/${kernelName};
+            echo '${config}' > kernels/${kernelName}/kernel.json;
+          '') cfg.kernels)}
+
+        mkdir $out
+        cp -r kernels $out
+      '';
+  };
+
+  notebookConfig = pkgs.writeText "jupyter_config.py" ''
+    c.NotebookApp.password = u'${cfg.password}'
+
+    ${cfg.notebookConfig}
+  '';
+
+in {
+
+  options.services.jupyter = {
+    enable = mkEnableOption "Jupyter development server";
+
+    ip = mkOption {
+      type = types.str;
+      default = "localhost";
+      description = ''
+        IP address Jupyter will be listening on.
+      '';
+    };
+
+    port = mkOption {
+      type = types.int;
+      default = 8888;
+      description = ''
+        Port number Jupyter will be listening on.
+      '';
+    };
+
+    notebookDir = mkOption {
+      type = types.str;
+      default = "~/";
+      description = ''
+        Root directory for notebooks.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "jupyter";
+      description = ''
+        Name of the user used to run the jupyter service.
+        For security reason, jupyter should really not be run as root.
+        If not set (jupyter), the service will create a jupyter user with appropriate settings.
+      '';
+      example = "aborsu";
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "jupyter";
+      description = ''
+        Name of the group used to run the jupyter service.
+        Use this if you want to create a group of users that are able to view the notebook directory's content.
+      '';
+      example = "users";
+    };
+
+    password = mkOption {
+      type = types.str;
+      description = ''
+        Password to use with notebook.
+        Can be generated using:
+          In [1]: from notebook.auth import passwd
+          In [2]: passwd('test')
+          Out[2]: 'sha1:1b961dc713fb:88483270a63e57d18d43cf337e629539de1436ba'
+      '';
+      example = "sha1:1b961dc713fb:88483270a63e57d18d43cf337e629539de1436ba";
+    };
+
+    notebookConfig = mkOption {
+      type = types.lines;
+      default = "";
+      description = ''
+        Raw jupyter config that will be appended to the generated config.
+      '';
+    };
+
+    kernels = mkOption {
+      type = types.attrsOf( types.submodule {
+        options = {
+          name = mkOption {
+            type = types.str;
+            default = "Python 3";
+            description = ''
+              Name to use for the kernel.
+            '';
+          };
+
+          env = mkOption {
+            type = types.package;
+            default = (pkgs.python36.withPackages (
+              pythonPackages: with pythonPackages; [
+                notebook
```

Comment:
`ipykernel` should be sufficient as the `notebook` should already be offered by this service

Reply:
Indeed, good catch